kv/kvcache/cache.go

```diff
@@ -400,43 +441,165 @@ func (c *Coherent) removeOldestCode(r *CoherentRoot) {
 		r.codeCache.Delete(e)
 	}
 }
-func (c *Coherent) add(k, v []byte, r *CoherentRoot, id ViewID) *Element {
+func (c *Coherent) add(k, v []byte, r *CoherentRoot, id uint64) *Element {
 	it := &Element{K: k, V: v}
 	replaced, _ := r.cache.ReplaceOrInsert(it)
-	if c.latestViewID != id {
+	if c.latestStateVersionID != id {
 		//fmt.Printf("add to non-last viewID: %d<%d\n", c.latestViewID, id)
 		return it
 	}
 	if replaced != nil {
 		c.stateEvict.Remove(replaced)
 	}
 	c.stateEvict.PushFront(it)
-	evict := c.stateEvict.Len() > c.cfg.KeysLimit
-	// Verify size not exceeded
-	if evict {
+
+	// clear down cache until size below the configured limit
+	for c.stateEvict.Size() > int(c.cfg.CacheSize.Bytes()) {
 		c.removeOldest(r)
 	}
+
 	return it
 }
-func (c *Coherent) addCode(k, v []byte, r *CoherentRoot, id ViewID) *Element {
+func (c *Coherent) addCode(k, v []byte, r *CoherentRoot, id uint64) *Element {
 	it := &Element{K: k, V: v}
 	replaced, _ := r.codeCache.ReplaceOrInsert(it)
-	if c.latestViewID != id {
+	if c.latestStateVersionID != id {
 		//fmt.Printf("add to non-last viewID: %d<%d\n", c.latestViewID, id)
 		return it
 	}
 	if replaced != nil {
 		c.codeEvict.Remove(replaced)
 	}
 	c.codeEvict.PushFront(it)
-	evict := c.codeEvict.Len() > c.cfg.CodeKeysLimit
-	// Verify size not exceeded
-	if evict {
+
+	for c.codeEvict.Size() > int(c.cfg.CodeCacheSize.Bytes()) {
 		c.removeOldestCode(r)
 	}
+
 	return it
 }
 
+func (c *Coherent) ValidateCurrentRoot(ctx context.Context, tx kv.Tx) (*CacheValidationResult, error) {
+
+	result := &CacheValidationResult{
+		Enabled:          true,
+		RequestCancelled: false,
+	}
+
+	select {
+	case <-ctx.Done():
+		result.RequestCancelled = true
+		return result, nil
+	default:
+	}
+
+	idBytes, err := tx.GetOne(kv.Sequence, kv.PlainStateVersion)
+	if err != nil {
+		return nil, err
+	}
+	stateID := binary.BigEndian.Uint64(idBytes)
+	result.LatestStateID = stateID
+
+	// if the latest view id in the cache is not the same as the tx or one below it
+	// then the cache will be a new one for the next call so return early
+	if stateID > c.latestStateVersionID {
+		result.LatestStateBehind = true
+		return result, nil
+	}
+
+	root := c.roots[c.latestStateVersionID]
+
+	// ensure the root is ready or wait and press on
+	select {
+	case <-root.ready:
+	case <-time.After(c.cfg.NewBlockWait):
+	}
+
+	// check context again after potentially waiting for root to be ready
+	select {
+	case <-ctx.Done():
+		result.RequestCancelled = true
+		return result, nil
+	default:
+	}
+
+	clearCache := false
+
+	compare := func(cache *btree.BTreeG[*Element], bucket string) (bool, [][]byte, error) {
+		keys := make([][]byte, 0)
+
+		for {
```

Comment:
1. likely we will need limit time of validation loop - to avoid "too long living read transactions" 
2. removekv already has protection against "too long living read transactions": `MaxTxTTL`. It will re-open read transaction after `MaxTxTTL` on server. And this check will start fail - because underlying transaction changed (will start show newer data after `MaxTxTTL` - inconsistent view).

Reply:
@AskAlexSharov I think we can see how long it typically takes to check the cache and perhaps it won't be a problem worth solving, I do not expect check cache be used by many users routinely

Reply:
ok.